lib/executor/constant_arrival_rate.go

```diff
@@ -216,37 +217,52 @@ func (car ConstantArrivalRate) Run(ctx context.Context, out chan<- stats.SampleC
 	}).Debug("Starting executor run...")
 
 	// Pre-allocate the VUs local shared buffer
-	vus := make(chan lib.VU, maxVUs)
+	activeVUs := make(chan lib.ActiveVU, maxVUs)
+	activeVUsCount := uint64(0)
 
-	initialisedVUs := uint64(0)
-	// Make sure we put planned and unplanned VUs back in the global
-	// buffer, and as an extra incentive, this replaces a waitgroup.
+	activeVUsWg := &sync.WaitGroup{}
+	defer activeVUsWg.Wait()
+	// Make sure we put planned and unplanned VUs back in the global buffer
 	defer func() {
-		// no need for atomics, since initialisedVUs is mutated only in the select{}
-		for i := uint64(0); i < initialisedVUs; i++ {
-			car.executionState.ReturnVU(<-vus, true)
+		currActiveVUs := atomic.LoadUint64(&activeVUsCount)
+		for i := uint64(0); i < currActiveVUs; i++ {
+			vu := <-activeVUs
```

Comment:
This executor currently hangs here (same for the variable-arrival-rate at the same place), because `activeVUsCount` is not decremented in the `DeactivateCallback`. Yet if I add that then `TestArrivalRateCancel` fails with `Run returned before all VU iterations were finished`.

I'm not clear why this happens, as `Run()` should wait for all `activeVUsWg.Done()` calls. Any ideas?

Reply:
This whole defer seems superfluous now? Aren't you returning all VUs (both pre-allocated and dynamically allocated ones) via their `DeactivateCallback` lambdas?

Reply:
That is, I think this hangs because you're trying to return VUs you've already returned

Reply:
@imiric, I'm not sure this covers everything, or that it doesn't have any bugs, but it seems to fix the hanging issue: https://github.com/loadimpact/k6/compare/fix/1007-1283-context-cancelled...1368-maybe-fixes?expand=1

Reply:
hmm nope, that code has bugs...

Reply:
Pushed another commit in that branch that I think fixes the issue properly. The logic I have for the end of the `Run()` method and all of the `defers` is like this:
1. (the `for { select {} }` bit) run iterations when we need to
2. (first `defer`, counting backwards) wait for all VUs to be returned to the local pool - they are returned there when they finish an iteration - we're not starting any new iterations by this time, so if all VUs are there, it means all iterations have completed
3. (`defer cancel()`) now that all VUs are no longer running anything, we don't need to wait for the full `gracefulStop` timeout, so we cancel that context, and this calls the `DeactivateCallback` of all `ActiveVU`s
4. (`defer activeVUsWg.Wait()`) wait until all active VUs have been deactivated and returned back to the global pool

Reply:
That makes sense, thanks for looking into it. I knew about the reverse `defer` call order, but I didn't consider it here at all :facepalm:

Would the same change be needed for variable-arrival-rate as well?

> This whole defer seems superfluous now?

Yeah, I first tried removing it, but ran into the same `TestArrivalRateCancel` failure, and the test seemed fine. Maybe a refactor that exposes a friendly API that simplifies how executors manage VUs and keep track of them would avoid this type of problems and boilerplate currently in all executors. After #1007 to be sure :)

Reply:
> Would the same change be needed for variable-arrival-rate as well?

I didn't look into that, but the VU management mechanics there should be very similar, so probably yes.

> Maybe a refactor that exposes a friendly API that simplifies how executors manage VUs and keep track of them would avoid this type of problems and boilerplate currently in all executors. After #1007 to be sure :)

Yeah, something like that is percolating at the back of my head, but it's definitely for after we release it.